### PR TITLE
[improvement](scanner) Remove the predicate that is always true for the segment

### PR DESCRIPTION
**Pull request by mrhhsg**
## Proposed changes

By utilizing the zonemap index of the segment, we can ascertain if a predicate is always true. For example, if the segment’s maximum value is 100 and the predicate is col < 101, then this predicate is always true for this segment.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 1]
```diff
@@ -157,6 +156,19 @@ Status Segment::new_iterator(SchemaSPtr schema, const StorageReadOptions& read_o
         iter->reset(new SegmentIterator(this->shared_from_this(), schema));
     }
 
+    auto pruned_predicates = read_options.column_predicates;
+    auto pruned = false;
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment.cpp:160]: add some regression test

File: be/src/olap/comparison_predicate.h [hunk 2]
```diff
@@ -174,6 +174,29 @@ class ComparisonPredicateBase : public ColumnPredicate {
         }
     }
 
+    bool is_always_true(const std::pair<WrapperField*, WrapperField*>& statistic) const override {
+        if (statistic.first->is_null()) {
```
Comment (yiguolei) [inline be/src/olap/comparison_predicate.h:178]: add  || statistic.second ->is null??

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 3]
```diff
@@ -338,6 +339,27 @@ bool ColumnReader::match_condition(const AndBlockColumnPredicate* col_predicates
                                      col_predicates);
 }
 
+bool ColumnReader::prune_predicates_by_zone_map(std::vector<ColumnPredicate*>& predicates,
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/column_reader.cpp:342]: add a config to disable this
and the prune should only be used in query, not in compaction.

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 4]
```diff
@@ -157,6 +156,19 @@ Status Segment::new_iterator(SchemaSPtr schema, const StorageReadOptions& read_o
...
+    for (auto& it : _column_readers) {
+        if (it.second->prune_predicates_by_zone_map(pruned_predicates, it.first)) {
+            pruned = true;
+        }
+    }
+
+    if (pruned) {
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment.cpp:167]: add a config to disable this
and the prune should only be used in query, not in compaction.

**Comment by mrhhsg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by jacktengg — APPROVED** [review]
LGTM

**Pull request merged**